### PR TITLE
Feat/add rules for returning parameters

### PR DESCRIPTION
**Pull request by L-M-Sherlock**
Context: https://github.com/ankitects/anki/issues/3094#issuecomment-2027764614

### CODE REVIEW

**Comment by dae** [comment]
I was thinking of doing a stable release later today or tomorrow. I'm ok with including this change in the stable build, if you are reasonably confident it won't cause a regression / doesn't need any testing.

**Comment by Expertium** [comment]
@L-M-Sherlock previously, you said "It's unnecessary to evaluate the default parameters. For all new users, the old_parameters is default_parameters". But that is true only for the very first optimization. It's not true for consecutive optimizations, so comparison with default parameters should be added.

**Comment by L-M-Sherlock** [comment]
> But that is true only for the very first optimization. It's not true for consecutive optimizations

Do you mean the optimized parameters become worse than the default parameters over time?

By the way, each evaluation slows down the optimization.

**Comment by Expertium** [comment]
> Do you mean the optimized parameters become worse than the default parameters over time?

Yes, it can happen sometimes. According to Rlz:

Size: 16384
Amt where 'default' does better than 'opt': 61 / 3299
Mean RMSE: 0.089 (default) vs 0.093 (opt)

Size: 32768
Amt where 'default' does better than 'opt': 34 / 2119
Mean RMSE: 0.107 (default) vs 0.112 (opt)

So even for very large sample sizes, default parameters are better 1-2% of the time.

**Comment by user1823** [comment]
> I'm ok with including this change in the stable build, if you are reasonably confident it won't cause a regression / doesn't need any testing.

I think that doing some more testing would be better. In addition, it would be better to gather feedback from users about things like whether this increases the optimization time significantly with their collection and hardware.

So, I would suggest releasing 24.04 as it is. This change can be included in the next release.

> So even for very large sample sizes, default parameters are better 1-2% of the time.

In such cases, would it be better to use the optimized values of first 4 parameters and default values of the rest rather than using the default values of all parameters?

**Comment by Expertium** [comment]
> In such cases, would it be better to use the optimized values of first 4 parameters and default values of the rest rather than using the default values of all parameters?

The rule compares all 3: default, pretrain and full. Of course, a situation where pretrain is better than full but worse than default would be exceptionally rare.

**Comment by user1823** [comment]
> Of course, a situation where pretrain is better than full but worse than default would be exceptionally rare.

Did Rlz find such cases during the analysis? If there were very few cases, we can skip comparing against the default parameters. This would atleast decrease the time taken for optimization. In addition, this can prevent the confusion arising from a sudden shift to the default parameters.

> This could cause confusion if the user is paying close attention to the values of parameters.

Not only if the user is paying attention to the parameters.
The parameters (especially the initial stability) affect the memory states significantly. So, jumping from optimized parameters to default parameters may significantly affect the next intervals of the cards.

**Comment by Expertium** [comment]
Size: 16384
Amt where 'default' does better than 'opt': 61 / 3299
Mean RMSE: 0.089 (default) vs 0.093 (opt)
Amt where 'opt' does better than 'default': 3238 / 3299
Mean RMSE: 0.078 (opt) vs 0.104 (default)
Amt where 'default' does better than 'pretrain': 338 / 3299
Mean RMSE: 0.094 (default) vs 0.096 (pretrain)
Amt where 'pretrain' does better than 'default': 2959 / 3299
Mean RMSE: 0.091 (pretrain) vs 0.105 (default)
Amt where 'opt' does better than 'pretrain': 3205 / 3299
Mean RMSE: 0.078 (opt) vs 0.092 (pretrain)
Amt where 'pretrain' does better than 'opt': 94 / 3299
Mean RMSE: 0.093 (pretrain) vs 0.094 (opt)

**Comment by user1823** [comment]
This doesn't answer the specific question. I asked for the cases where the RMSE is in the order:
 
def < pretrain < opt

Answering the question requires looking at the raw data. This summary is not that useful.

Also, what about allowing small regressions in RMSE, considering that RMSE may not be the "perfect" metric?

**Comment by Expertium** [comment]
@RlzHi could you please analyze all possible permutations of these 3 losses, to see how often certain permutations occur?

For example, how often default < pretrain < opt occurs, or how often pretrain < default < opt occurs

**Comment by L-M-Sherlock** [comment]
I find a new problem: without the minimum limit, the user will only get this error:

image

Because the first review entry of new cards are removed, so the number of reviews for optimization is zero.

**Comment by Expertium** [comment]
Just use the default parameters

**Comment by L-M-Sherlock** [comment]
So, should we remove this error?

**Comment by Expertium** [comment]
Yes

**Comment by L-M-Sherlock** [comment]
But how should the `evaluate` feature behave if the number of reviews is zero?

**Comment by Expertium** [comment]
Evaluate default parameters, since that's what will be used

**Comment by L-M-Sherlock** [comment]
> Evaluate default parameters, since that's what will be used

But the size of testset is zero. We can't evaluate the parameters in an empty testset.

**Comment by user1823** [comment]
There are no reviews to evaluate the parameters. So, obviously, the error should be shown.

**Comment by Expertium** [comment]
Oh, that's what you meant. My bad. Yeah, then an error should be shown.

**Comment by L-M-Sherlock** [comment]
OK, this PR will be merged after @dae or @asukaminato0721 reviews it.

**Comment by Expertium** [comment]
> Did Rlz find such cases during the analysis? If there were very few cases, we can skip comparing against the default parameters. This would atleast decrease the time taken for optimization. In addition, this can prevent the confusion arising from a sudden shift to the default parameters.

From a purely "decrease RMSE as much as possible" point of view, this rule is as good as we can get. Any further improvement would be extremely minor. But it's true that we also need to avoid confusion among users. So perhaps skipping comparison with the default parameters would be better, even if it goes against minimizing RMSE at any cost. Initially, I was leaning more towards not skipping it, but now I think it's better to skip it.
I'm typing this out for Dae as well, so that he can understand what's going on.

**Review by asukaminato0721 — APPROVED** [review]

**Review by Expertium — APPROVED** [review]

File: src/training.rs [hunk 1]
```diff
@@ -490,10 +495,32 @@ mod tests {
...
+        let pretrained_rmse = pretrained_fsrs
+            .evaluate(testset.clone(), |_| true)
+            .unwrap()
+            .rmse_bins;
+        let default_fsrs = FSRS::new(Some(&[])).unwrap();
+        let default_rmse = default_fsrs
+            .evaluate(testset.clone(), |_| true)
+            .unwrap()
+            .rmse_bins;
+        let (best_parameters, best_rmse) = if pretrained_rmse < default_rmse {
```
Comment (user1823) [inline src/training.rs:518]: I think that these tests are no longer needed.

Comment (Expertium) [reply src/training.rs:518]: Good catch

Comment (L-M-Sherlock) [reply src/training.rs:518]: It is just used for test.

Comment (L-M-Sherlock) [reply src/training.rs]: Any other feedback?

Comment (user1823) [reply src/training.rs]: > Any other feedback?

Nothing more from me.

**Pull request merged**